third_party/move/move-bytecode-verifier/src/signature_v2.rs

```diff
@@ -0,0 +1,984 @@
+                    )?);
+                }
+
+                *entry.insert(self.constraints.alloc(constraints))
+            },
+        };
+        Ok(r)
+    }
+
+    /// Checks if all field instantiations are well-formed, in a context-less fashion.
+    ///
+    /// Time complexity: `O(total_size_of_all_type_args_in_all_instantiations)`
+    fn verify_field_instantiations(&self) -> PartialVMResult<()> {
+        for field_inst_idx in 0..self.resolver.field_instantiations().unwrap().len() {
+            self.verify_field_instantiation(FieldInstantiationIndex(field_inst_idx as u16))?;
+        }
+        Ok(())
+    }
+
+    /// Checks if a function handle is well-formed.
+    ///
+    /// A function handle is well-formed if all parameter and return types are well-formed, with references
+    /// being allowed.
+    ///
+    /// Time complexity: `O(total_size_of_all_types)` if not cached.
+    fn verify_function_handle(&self, fh: &FunctionHandle) -> PartialVMResult<()> {
+        let ability_context = BitsetTypeParameterConstraints::from(fh.type_parameters.as_slice());
+        self.verify_signature_in_context(&ability_context, fh.parameters)?;
+        self.verify_signature_in_context(&ability_context, fh.return_)?;
+        Ok(())
+    }
+
+    /// Checks if all function handles are well-formed.
+    fn verify_function_handles(&self) -> PartialVMResult<()> {
+        for (idx, fh) in self.resolver.function_handles().iter().enumerate() {
+            self.verify_function_handle(fh)
+                .map_err(|err| err.at_index(IndexKind::FunctionHandle, idx as u16))?;
+        }
+        Ok(())
+    }
+
+    /// Checks if a code unit is well-formed.
+    ///
+    /// A code unit is well-formed if
+    /// - The locals are well-formed within the context. (References are allowed.)
+    /// - All instantiations (function, struct, field & vector) are well-formed within the context and do not
+    ///   contain references.
+    ///
+    /// Time complexity: `O(num_of_ty_params * num_of_instantiations)`,
+    ///                  assuming that the verification results for the instantiations have been cached.
+    fn verify_code(&self, ability_context: &[AbilitySet], code: &CodeUnit) -> PartialVMResult<()> {
+        use Bytecode::*;
+
+        let ability_context = BitsetTypeParameterConstraints::from(ability_context);
+
+        self.verify_signature_in_context(&ability_context, code.locals)
+            .map_err(|err| err.at_index(IndexKind::Signature, code.locals.0))?;
+
+        // Local caches to avoid re-verifying identical instantiations within the same context.
+        let mut checked_func_insts = BTreeMap::<FunctionInstantiationIndex, ()>::new();
+        let mut checked_struct_def_insts = BTreeMap::<StructDefInstantiationIndex, ()>::new();
+        let mut checked_struct_def_insts_with_key =
+            BTreeMap::<StructDefInstantiationIndex, ()>::new();
+        let mut checked_vec_insts = BTreeMap::<SignatureIndex, ()>::new();
```

Comment:
nit: I think what we want is a BtreeSet? also I don't understand much about the index range, but if it's short, maybe a bitmask is more efficient?